### PR TITLE
CRM-20332 Added support for third column of actions list to be modifi…

### PR DESCRIPTION
**Pull request by Edzelopez**
…ed by hook summaryActions

----------------------------------------
* CRM-20332: Refactor summaryActions hook to allow user-defined actions in third column
  https://issues.civicrm.org/jira/browse/CRM-20332

### CODE REVIEW

**Comment by Edzelopez** [comment]
@monishdeb could you please QA this?

**Comment by monishdeb** [comment]
sure

**Comment by eileenmcnaughton** [comment]
@monishdeb you might be able to trade QA on this with @jitendrapurohit who has a few things awaiting QA I believe.

Based on a quick look at the code I do think we should try not to be deriving the urls from the smarty variables -ie 

$userRecordUrl = CRM_Core_Config::singleton()->userSystem->getUserRecordUrl($cid);
rather than
 if ($userRecordUrl = CRM_Core_Smarty::singleton()->get_template_vars('userRecordUrl')) {
although there might be some issues to resolve.

Question - won't this break things for people who are already overriding existing actions? Perhaps instead of putting more layers into the menu you need to add some fields to the items in the array - since I assume you are using the layers to denote which column to use?

**Comment by Edzelopez** [comment]
@eileenmcnaughton I have now changed the function to prevent modification of the original array.

@monishdeb could you please QA or assign to someone appropriate for QA?

Thanks!

**Comment by monishdeb** [comment]
@jitendrapurohit could you please QA this PR?

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -3006,6 +3006,80 @@ public static function contextMenu($contactId = NULL) {
...
+    if (CRM_Core_Permission::check('access Contact Dashboard')) {
+      $menu['otherActions']['dashboard'] = array(
+        'title' => ts('Contact Dashboard'),
+        'description' => ts('Contact Dashboard'),
+        'weight' => 15,
+        'ref' => 'crm-contact-dashboard',
+        'key' => 'dashboard',
+        'tab' => 'dashboard',
+        'class' => 'dashboard',
+        'href' => CRM_Utils_System::url('civicrm/user',
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php]: I realise this drupal-specific method was already in use, but, could you add a todo suggesting it be derived in a CMS-based way similar to CRM_Core_Config::singleton()->userSystem->getUserRecordUrl($contactId), to make it clear that it's a short-cut

**Comment by monishdeb** [comment]
@eileenmcnaughton I have updated the PR, ready for QA

**Review by jitendrapurohit — APPROVED** [review]

**Comment by jitendrapurohit** [comment]
Checked the hook by inserting menu from `$actions['otherActions']['casework'] = array(...)` and confirmed that it gets inserted to the third column correctly.

**Comment by monishdeb** [comment]
@eileenmcnaughton can you merge the PR as @jitendrapurohit has approved the changes. Thanks

**Comment by davejenx** [comment]
This PR appears to cause CRM-20467 PHP notices on contact search for ACL'd user due to use of undefined $values in checkUserMenuPermissions().

**Comment by Edzelopez** [comment]
Sorry about that @davejenx, it was a typo that was introduced in an effort to generalize code which has been fixed by @monishdeb [here](https://github.com/civicrm/civicrm-core/pull/10243/commits/c2463857308dd897978f84c41bab7e0e788813f9)

**Comment by davejenx** [comment]
@Edzelopez Thanks, confirmed that #10243 fixes it.

**Pull request merged**